packages/framework/deprecated-agent-scheduler-container-runtime/src/agentSchedulerBaseContainerRuntimeFactory.ts

```diff
@@ -0,0 +1,113 @@
+    buildRuntimeRequestHandler,
+    innerRequestHandler,
+} from "@fluidframework/request-handler";
+import {
+    IFluidDataStoreRegistry,
+    IProvideFluidDataStoreRegistry,
+    NamedFluidDataStoreRegistryEntries,
+} from "@fluidframework/runtime-definitions";
+import { DependencyContainer, DependencyContainerRegistry } from "@fluidframework/synthesize";
+import {
+    makeContainerRuntimeWithAgentScheduler,
+} from "./makeContainerRuntimeWithAgentScheduler";
+
+/**
+ * BaseContainerRuntimeFactory produces container runtimes with a given data store and service registry, as well as
+ * given request handlers.  It can be subclassed to implement a first-time initialization procedure for the containers
+ * it creates.
+ * @deprecated Only use if your scenario requires backwards compatibility with documents that were produced before
+ * AgentScheduler was removed from ContainerRuntime.
+ */
+export class AgentSchedulerBaseContainerRuntimeFactory implements
+    IProvideFluidDataStoreRegistry,
+    IRuntimeFactory {
+    public get IFluidDataStoreRegistry() { return this.registry; }
+    public get IRuntimeFactory() { return this; }
+    private readonly registry: IFluidDataStoreRegistry;
+
+    /**
+     * @param registryEntries - The data store registry for containers produced
+     * @param serviceRegistry - The service registry for containers produced
+     * @param requestHandlers - Request handlers for containers produced
+     * @param runtimeOptions - The runtime options passed to the ContainerRuntime when instantiating it
+     * @deprecated Only use if your scenario requires backwards compatibility with documents that were produced before
+     * AgentScheduler was removed from ContainerRuntime.
+     */
+    constructor(
+        private readonly registryEntries: NamedFluidDataStoreRegistryEntries,
+        private readonly providerEntries: DependencyContainerRegistry = [],
+        private readonly requestHandlers: RuntimeRequestHandler[] = [],
+        private readonly runtimeOptions?: IContainerRuntimeOptions,
+    ) {
+        this.registry = new FluidDataStoreRegistry(registryEntries);
+    }
+
+    /**
+     * {@inheritDoc @fluidframework/container-definitions#IRuntimeFactory.instantiateRuntime}
+     */
+    public async instantiateRuntime(
+        context: IContainerContext,
+    ): Promise<IRuntime> {
+        const parentDependencyContainer = context.scope.IFluidDependencySynthesizer;
+        const dc = new DependencyContainer(parentDependencyContainer);
+        for (const entry of Array.from(this.providerEntries)) {
+            dc.register(entry.type, entry.provider);
+        }
+
+        // Create a scope object that passes through everything except for IFluidDependencySynthesizer
+        // which we will replace with the new one we just created.
+        const scope: any = context.scope;
+        scope.IFluidDependencySynthesizer = dc;
+
+        const runtime = await makeContainerRuntimeWithAgentScheduler(
```

Comment:
This class really does not have anything to do with agent scheduler other than this call.
Which feels like we can use mixin model not to duplicate code (here, same goes about AgentSchedulerContainerRuntimeFactoryWithDefaultDataStore).


I.e. something like that:

```javascript
// Same as today's BaseContainerRuntimeFactory, with whiteout dependency logic - really only skeleton, no customizations! Ideally even things like innerRequestHandler are not here!
class RealBaseContainerRuntimeFactory implements ... {
...
    public async instantiateRuntime(
        context: IContainerContext,
    ): Promise<IRuntime> {
        const runtime = await ContainerRuntime.load(
            context,
            this.registryEntries,
            buildRuntimeRequestHandler(
                ...this.requestHandlers,
                innerRequestHandler),
            this.runtimeOptions,
            scope);

        if (!runtime.existing) {
            // If it's the first time through.
            await this.containerInitializingFirstTime(runtime);
        }

        // This always gets called at the end of initialize on first time or from existing.
        await this.containerHasInitialized(runtime);

        return runtime;
}
...
}
```

And them BaseContainerRuntimeFactory becomes a mixin:
```javascript
export const createBaseContainerRuntimeFactory = (
    Base: typeof RealBaseContainerRuntimeFactory  = RealBaseContainerRuntimeFactory,
) => class Derived extends Base {

    public async instantiateRuntime(
        context: IContainerContext,
    ): Promise<IRuntime> {
      ... // DependencyContainer specific code
      super.instantiateRuntime();
      ... // DependencyContainer specific code
}

} as typeof RealBaseContainerRuntimeFactory;
```

And now there is no need for AgentSchedulerBaseContainerRuntimeFactory 
 & AgentSchedulerContainerRuntimeFactoryWithDefaultDataStore. We just need a mixin that adds agent scheduler. And that class can be inserted into inheritance chain in any order, using any base class.

I know it's less readable, but it is so much more powerful! You can compose and change behavior however you want.
For example, I can plug in my own base that does not use ContainerRuntime at all (i.e. has alternative implementation of IContainerRuntime). Or I can have derived class from ContainerRuntime with modified behavior and overwrite certain things (like handling of dirty flag).

Ideally in all of this the real base class has no logic. I.e. its as barebones as possible, ensuring that callers can actually get any implementation they want out of it by composing such layers.